frame/dapps-staking/src/fix.rs

```diff
@@ -0,0 +1,175 @@
+use super::*;
+use codec::Decode;
+
+pub mod restake_fix {
+
+    use super::*;
+    use codec::Encode;
+    use frame_support::log;
+    use frame_support::{
+        storage::generator::StorageMap, storage::PrefixIterator, traits::Get, weights::Weight,
+    };
+    use sp_runtime::traits::Saturating;
+    use sp_std::collections::btree_map::BTreeMap;
+
+    // Temp struct of corrected ContractStakeInfo records with progress data
+    #[derive(Clone, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo, Default)]
+    pub struct RestakeFix<Balance: HasCompact> {
+        // store last processed staker from Ledger
+        last_fully_processed_staker: Option<Vec<u8>>,
+        // store last processed staker's dapp
+        last_processed_staker_contract: Option<Vec<u8>>,
+        // accumulate ContractStakeInfo by contract address
+        contract_staking_info: BTreeMap<Vec<u8>, ContractStakeInfo<Balance>>,
+        // flip to true when all Ledger data has been processed
+        all_stakers_processed: bool,
+    }
+
+    pub fn restake_fix_migration<T: Config>(weight_limit: Weight) -> Weight {
+        log::info!("Executing a step of restake fix migration.");
+        let mut restake_fix = RestakeFixAccumulator::<T>::get();
+        // one read for RestakeFixAccumulator
+        let mut consumed_weight = T::DbWeight::get().reads(1);
+        // still stakers left to process
+        if !restake_fix.all_stakers_processed {
+            log::info!("Accumulating staker info.");
+            // take iter from start or last processed
+            let staker_iter = if let Some(last_processed_staker) =
+                restake_fix.last_fully_processed_staker.clone()
+            {
+                Ledger::<T>::iter_keys_from(last_processed_staker)
+            } else {
+                Ledger::<T>::iter_keys()
+            };
+            // star iterating, label to be able to break out from inner loop
+            'staker_iter: for staker in staker_iter {
+                // every iteration adds one more read
+                consumed_weight = consumed_weight.saturating_add(T::DbWeight::get().reads(1));
+                // take iter for inner loop from start or last processed contract
+                let contract_iter = if let Some(last_processed_contract) =
+                    restake_fix.last_processed_staker_contract.clone()
+                {
+                    GeneralStakerInfo::<T>::iter_prefix_from(&staker, last_processed_contract)
+                } else {
+                    GeneralStakerInfo::<T>::iter_prefix(&staker)
+                };
```

Comment:
You can remove this and process the entire batch for each staker - even if you go over the weight limit, it will be negligible.

This will also simplify your logic a lot.